### PR TITLE
Add generic function to determine if URL is a store page

### LINKED ISSUE #45110: [Launch Your Store]  Add generic function to determine if URL is a store page
### Issue description

Add a generic function for determining if URL is a store page so that we can keep store pages in “Coming Soon” mode while keeping the rest of the site live.

### Accept criteria

- function should return true if URL is a store page, return false, otherwise

Store pages are defined as:

* Shop Page
* Cart Page
* Checkout Page
* My Account Page
* Terms and Conditions Page
* Privacy Policy Page

(the above page IDs can be determined from WooCommerce Settings)

Additionally, the following autogenerated pages should be included:
* Product Categories (taxonomy=product_cat)
* Product Tags (taxonomy=product_tag)

### PR DESCRIPTION
**Pull request by chihsuan**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/45110.

The function `is_store_page` is added to `WCAdminHelper` and is used to determine if a URL is a store page.

Store pages are defined as:

- Shop Page
- Cart Page
- Checkout Page
- My Account Page
- Terms and Conditions Page
- Privacy Policy Page

Additionally, the following autogenerated pages should be included:

- Product
- Product Categories (taxonomy=product_cat)
- Product Tags (taxonomy=product_tag)

The function ignores the domain and protocol of the URL and only checks the path and query string so that users can only pass `$_SERVER['REQUEST_URI']` to the function for checking if the current page is a store page.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a fresh site
2. Install `WooCommerce Smooth Generator` and `Code Snippets` plugins
3. Either skip or complete Core profiler
4. Go to `/wp-admin/tools.php?page=smoothgenerator` > Generate 10 products  OR run `wp wc generate products 10`
5. Go to `/wp-admin/admin.php?page=snippets`
6. Add new Snippet

```php
use Automattic\WooCommerce\Admin\WCAdminHelper;

function maybe_add_banner_to_footer() {
	if ( wp_doing_ajax() || wp_doing_cron() || WC()->is_rest_api_request() ) {
		return;
	}

	$is_store = WCAdminHelper::is_store_page();
	if ( $is_store ) {
    	echo '<div style="position: fixed; bottom: 0; background-color: red; color: white; width: 100%; text-align: center;">This is a store page!</div>';
	}
}

add_action('wp_footer', 'maybe_add_banner_to_footer');
```

7.  Select `Only run on site front-end` and activate
8. Go to frontend
9. Confirm the `This is a store page!` red banner is displayed on the store pages (Shop, Cart, Checkout, My Account, Terms and Conditions, Privacy Policy, Product, Product Categories, Product Tags) and not displayed on other pages (Home, Sample Page, etc.)
10. Go to `/wp-admin/admin.php?page=wc-settings&tab=advanced`
11. Change `Terms and conditions` page to Sample Page
12. Go to Sample Page
13. Confirm the banner is displayed on the Sample Page
14. Go to `/wp-admin/options-permalink.php`
15. Change Product permalinks to `Shop base with category`
16. Change Product category base to `test-category`
17. Go to a product page
18. Confirm the banner is still displayed on the product page
19. Go to a product category page

Screenshot 2024-03-05 at 11 32 37

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Admin/WCAdminHelper.php [hunk 1]
```diff
@@ -122,4 +122,115 @@ public static function is_site_fresh() {
...
+	 * 3. Remove leading and trailing slashes.
+	 *
+	 * For example:
+	 *
+	 * - https://example.com/wordpress/shop/uncategorized/test/?add-to-cart=123 => shop/uncategorized/test/?add-to-cart=123
+	 *
+	 * @param string $url URL to normalize.
+	 */
+	private static function get_normalized_url_path( $url ) {
+		$quey            = wp_parse_url( $url, PHP_URL_QUERY );
```
Comment (rjchow) [inline plugins/woocommerce/src/Admin/WCAdminHelper.php]:
```suggestion
		$query            = wp_parse_url( $url, PHP_URL_QUERY );
```

File: plugins/woocommerce/src/Admin/WCAdminHelper.php [hunk 2]
```diff
@@ -122,4 +122,115 @@ public static function is_site_fresh() {
...
+	public static function is_store_page( $url = '' ) {
+		$url = $url ? $url : esc_url_raw( wp_unslash( $_SERVER['REQUEST_URI'] ?? '' ) );
+
+		if ( ! $url ) {
+			return false;
+		}
+		$normalized_path = self::get_normalized_url_path( $url );
+
+		// WC store pages.
+		$store_pages = array(
```
Comment (rjchow) [inline plugins/woocommerce/src/Admin/WCAdminHelper.php:155]: I think we can add a filter here for extensibility, what do you think?

**Review by psealock — COMMENTED** [review]
Code looks good @chihsuan. What about a page with a WooCommerce block, such as the All Product block. Would that count as a store page?

**Comment by chihsuan** [comment]
> What about a page with a WooCommerce block, such as the All Product block. Would that count as a store page?

Good question. 👍  I think it would but I'm not very sure. @adrianduffell, what do you think? This makes me think that we may need to add a filter so that 3rd party developers can add their own store pages to the list of store pages.

**Comment by adrianduffell** [comment]
> What about a page with a WooCommerce block, such as the All Product block. Would that count as a store page?

It's a good point. I think we could tackle this after the MVP. For now, the "restrict to store pages only" setting will link to external documentation that lists the pages which are included. We could add a note in the documentation about blocks not being supported yet.

We would need to come up with a strategy to detect pages with a WooCommerce block without a performance hit on every front-end request. Perhaps we could build an index of block pages each time a page gets saved. The filter also sounds like a good idea 👍

File: plugins/woocommerce/src/Admin/WCAdminHelper.php [hunk 3]
```diff
@@ -122,4 +122,115 @@ public static function is_site_fresh() {
...
+				continue;
+			}
+
+			if ( 0 === strpos( $normalized_path, self::get_normalized_url_path( $permalink ) ) ) {
+				return true;
+			}
+		}
+
+		// Check product, category and tag pages.
+		$permalink_structure = wc_get_permalink_structure();
```
Comment (rjchow) [inline plugins/woocommerce/src/Admin/WCAdminHelper.php:188]: not sure how big of an impact it is on performance but there's a get_option() call in `wc_get_permalink_structure()`

File: plugins/woocommerce/src/Admin/WCAdminHelper.php [hunk 4]
```diff
@@ -122,4 +122,115 @@ public static function is_site_fresh() {
...
+		}
+		$normalized_path = self::get_normalized_url_path( $url );
+
+		// WC store pages.
+		$store_pages = array(
+			'myaccount' => wc_get_page_id( 'myaccount' ),
+			'shop'      => wc_get_page_id( 'shop' ),
+			'cart'      => wc_get_page_id( 'cart' ),
+			'checkout'  => wc_get_page_id( 'checkout' ),
+			'privacy'   => wc_privacy_policy_page_id(),
```
Comment (rjchow) [inline plugins/woocommerce/src/Admin/WCAdminHelper.php:160]: not sure how big of an impact it is on performance but there's a get_option() call in wc_privacy_policy_page_id()

**Review by rjchow — COMMENTED** [review]
I left some comments on possible performance implications, not too sure if they are impactful

**Review by psealock — APPROVED** [review]
This is testing well, nice work @chihsuan you've covered a lot of scenarios.

I couldn't spot any issues with the code (other than @rjchow's comments) 🚀

File: plugins/woocommerce/src/Admin/WCAdminHelper.php (hunk 1 above)
Comment (chihsuan) [reply plugins/woocommerce/src/Admin/WCAdminHelper.php]: Fixed in d84d9f20d955c19bfc6f5409993837cb43160d2f

File: plugins/woocommerce/src/Admin/WCAdminHelper.php (hunk 2 above)
Comment (chihsuan) [reply plugins/woocommerce/src/Admin/WCAdminHelper.php:155]: Sounds good. Added a filter here https://github.com/woocommerce/woocommerce/pull/45299/commits/095c96345e70061a16afe801489bb10c56698247 🙂

File: plugins/woocommerce/src/Admin/WCAdminHelper.php (hunk 3 above)
Comment (chihsuan) [reply plugins/woocommerce/src/Admin/WCAdminHelper.php:188]: That's a good point. I think it's a minor impact. If it becomes a problem, we can optimize it later.

Comment (chihsuan) [reply plugins/woocommerce/src/Admin/WCAdminHelper.php:188]: From ChatGPT:

The impact of calling `wc_get_permalink_structure()` on performance depends on various factors such as server configuration, database size, and the context in which the function is being called.

1. **Caching**: If the result of `wc_get_permalink_structure()` is cached, subsequent calls to the function within the same request will have minimal impact on performance. WordPress itself caches the permalink structure, so in most cases, calling this function won't cause a significant performance hit.

2. **Database**: The function involves querying the database to retrieve the permalink structure. The impact may be noticeable on large or busy websites with heavy database loads. However, for most websites, especially those with caching mechanisms in place, the impact should be minimal.

3. **Context**: If the function is called frequently or in a critical performance path, such as within a loop that executes many times, the cumulative impact may become more significant. In such cases, it's advisable to call the function sparingly or optimize the code structure to minimize redundant calls.

4. **Server Resources**: The impact also depends on the server resources available. On high-traffic websites or under heavy server load, the impact may be more noticeable compared to low-traffic websites or under light server load.

Overall, calling `wc_get_permalink_structure()` typically has a minor impact on performance in most WordPress environments, especially when cached properly. However, it's essential to consider the specific context and optimize code accordingly if performance issues arise.

**Review by rjchow — APPROVED** [review]
Thanks for addressing the feedback!

**Pull request merged**